File: pulse-react-native-otel/android/src/main/java/com/pulsereactnativeotel/ReactNativeScreenAttributesSpanProcessor.kt
package com.pulsereactnativeotel

import io.opentelemetry.android.common.RumConstants
import io.opentelemetry.context.Context
import io.opentelemetry.sdk.trace.ReadWriteSpan
import io.opentelemetry.sdk.trace.ReadableSpan
import io.opentelemetry.sdk.trace.SpanProcessor

/**
 * Span processor that overrides screen.name with React Native screen name.
 */
internal class ReactNativeScreenAttributesSpanProcessor : SpanProcessor {
    override fun onStart(
        parentContext: Context,
        span: ReadWriteSpan,
    ) {
        ReactNativeScreenNameTracker.getCurrentScreenName()?.let { screenName ->
            span.setAttribute(RumConstants.SCREEN_NAME_KEY, screenName)
        }
    }

    override fun isStartRequired(): Boolean = true

    override fun onEnd(span: ReadableSpan) {
        // No-op
    }

    override fun isEndRequired(): Boolean = false
}

File: pulse-react-native-otel/ios/ReactNativeScreenAttributesSpanProcessor.swift
import Foundation
import OpenTelemetrySdk
import OpenTelemetryApi
import PulseKit

/**
 * Span processor that overrides screen.name with React Native screen name.
 */
internal class ReactNativeScreenAttributesSpanProcessor: SpanProcessor {
    var isStartRequired: Bool = true
    var isEndRequired: Bool = false
    
    func onStart(parentContext: SpanContext?, span: ReadableSpan) {
        if let screenName = ReactNativeScreenNameTracker.getCurrentScreenName() {
            span.setAttribute(key: PulseAttributes.screenName, value: AttributeValue.string(screenName))
        }
        if let previousScreenName = ReactNativeScreenNameTracker.getPreviousScreenName() {
            span.setAttribute(key: PulseAttributes.lastScreenName, value: AttributeValue.string(previousScreenName))
        }
    }
    
    func onEnd(span: any OpenTelemetrySdk.ReadableSpan) {
        // No-op
    }
    
    func shutdown(explicitTimeout: TimeInterval?) {
    }
    
    func forceFlush(timeout: TimeInterval?) {
    }
}

